### PR TITLE
utils/pypi: allow updating PyPI resources for non-PyPI formulae

### PR DESCRIPTION
**Pull request by cho-m**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

* Allow passing in `--extra-packages` with `--ignore-non-pypi-packages` to update PyPI resources when main formula is not a PyPI package
* Allow updating PyPI resources when main formula only has wheels

-----

Still have to fully test out behavior, e.g. scenarios where `pypi_info` was expected to return string, parts of code that expected `main_package` to be non-nil, etc.

Example scenarios I was looking at:

1. ` brew update-python-resources bazarr --extra-packages='lxml,webrtcvad-wheels' --ignore-non-pypi-packages`

2. `brew update-python-resources arcade-learning-environment --package-name ale_py`

    Currently fails with 
    ```
    Error: "ale_py" is not available on PyPI.
    ```
    
    With changes, it can pick up resource updates even though PyPI package only has wheels:
    ```
    ==> Retrieving PyPI dependencies for "ale_py"...
    ==> Excluding "ale-py==0.8.0"
    ==> Getting PyPI info for "importlib-resources==5.10.0"
    ==> Getting PyPI info for "numpy==1.23.4"
    ==> Updating resource blocks
    ```

### CODE REVIEW

File: Library/Homebrew/utils/pypi.rb [hunk 1]
```diff
@@ -93,9 +96,9 @@ def same_package?(other)
     end
 
     # Compare only names so we can use .include? and .uniq on a Package array
-    sig { params(other: Package).returns(T::Boolean) }
+    sig { params(other: T.untyped).returns(T::Boolean) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/pypi.rb:99]: Feels like it might be nicer to keep this typed?

File: Library/Homebrew/utils/pypi.rb [hunk 2]
```diff
@@ -167,30 +170,34 @@ def update_python_resources!(formula, version: nil, package_name: nil, extra_pac
...
-
-      odie "\"#{main_package}\" is not available on PyPI."
+      if ignore_non_pypi_packages
+        return if extra_packages.blank?
+      else
+        odie <<~EOS
+          Could not infer PyPI package name from URL:
+            #{Formatter.url(formula.stable.url)}
+        EOS
+      end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/pypi.rb:182]:
```suggestion
      if ignore_non_pypi_packages
        return if extra_packages.blank?
      end
      
      odie <<~EOS
        Could not infer PyPI package name from URL:
          #{Formatter.url(formula.stable.url)}
      EOS
```

Nice to remove/reduce this identation when possible on `return` or `odie`. Can do the same below too.

Comment (cho-m) [reply Library/Homebrew/utils/pypi.rb:182]: Want to allow continuing when `ignore_non_pypi_packages && extra_packages.present?`, i.e.:
- `ignore_non_pypi_packages && extra_packages.blank?` --> return
- `ignore_non_pypi_packages && extra_packages.present?` --> only update resources for extra_packages
- `!ignore_non_pypi_packages` --> odie

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/pypi.rb:182]: Ok, I see.

File: Library/Homebrew/utils/pypi.rb (hunk 1 above)
Comment (cho-m) [reply Library/Homebrew/utils/pypi.rb:99]: I'm not familiar with Sorbet, but part of code I added ended up with `nil` comparison (as `main_package` ended up `nil`). I do think `brew typecheck` still passed, but was thinking this needed at least `T.nilable(Package)` and then just ended up allowing any comparison.

https://github.com/Homebrew/brew/blob/06b7573e0b87eeb67907658d0004548b7c72c0d3/Library/Homebrew/utils/pypi.rb#L292

Still thinking on best way to refactor code to help create a non-pypi `Package` that can be used to perform the `#{package_name}@#{formula.stable.url}` logic that I mentioned in latest comment.

**Comment by woodruffw** [comment]
I believe this needs to be addressed upstream by `ale_py`: while not technically forbidden, wheel-only packages on PyPI are murky in terms of licensing (since the wheels might contain fully compiled code, and therefore not comply with their associated open source license requiring the index to host the source).

In other words, I *think* `ale_py` should be providing an sdist here, which would resolve the problem. More generally I think these changes make sense, though (even though non-PyPI dependencies in Python packages are discouraged)!

**Comment by MikeMcQuaid** [comment]
Let's reopen this when it's ready to ship, thanks 🙇🏻

**Pull request closed without merging**